BigInt.c

```diff
@@ -1,69 +1,157 @@
 #include <assert.h>
 #include <math.h>
+#include <errno.h>
+#include <limits.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 
 #include "BigInt.h"
+#include "safe_math_impl.h"
 
 #define MAX(x, y) ((x) > (y) ? (x) : (y))
 
+#if UNIT_MAX >> 32 == 0
+#	define check_add_int_int check_add_int32_int32
+#	define check_add_uint_uint check_add_uint32_uint32
+#	define check_mul_int_int check_mul_int32_int32
+#	define check_mul_uint_uint check_mul_uint32_uint32
+#else
+#	if UNIT_MAX >> 64 == 0
+#		define check_add_int_int check_add_int64_int64
+#		define check_add_uint_uint check_add_uint64_uint64
+#		define check_mul_int_int check_mul_int64_int64
+#		define check_mul_uint_uint check_mul_uint64_uint64
+#	else
+#		error unsupported integer size
+#	endif
+#endif
+
+#ifndef BIGINT_REDZONE
+#define BIGINT_REDZONE 0
+#endif//BIGINT_REDZONE
+
+#if BIGINT_REDZONE
+unsigned char* malloc_digits(unsigned int num_digits) {
+    unsigned int bytes;
+    if(
+        !check_mul_uint_uint(num_digits, sizeof(unsigned char), &bytes)
+        || !check_add_uint_uint(bytes, BIGINT_REDZONE * 2, &bytes)
+    ) {
+        errno = ENOMEM;
+        return NULL;
+    }
+    unsigned char* p = malloc(bytes);
+    memset(p, 0x42, bytes);
+    return p + BIGINT_REDZONE;
+}
+
+unsigned char* okay_digits(unsigned char* digits, unsigned int num_digits) {
```

Comment:
This is clever. Maybe worth a brief comment at the top of the redzone block outlining the checking scheme?